### PR TITLE
Accelerate first token gen with BF16-gemm MHA and concat-Silu MLP

### PR DESCRIPTION
**Pull request by abenmao**
1. BF16-based flash attention. Enabled when prompt_len >=1024
2. concat-silu gate+up proj
3. Added intel-mkl library in cmake
4. In addition, an environment variable ENABLE_CBLAS_MLP can be set to 1 (change downProj with cblas kernel) when prompt len ​​>= 2048. The default is 0.

### CODE REVIEW

File: src/layers/attention.h [hunk 1]
```diff
@@ -878,7 +914,8 @@ class Attention {
     }
 
     virtual const float *getMask(const float *attnMask, int bId, int hId, int srcLen, int tgtLen) {
-        return attnMask + bId * srcLen * tgtLen;
+        return attnMask;
```
Comment (pujiang2018) [inline src/layers/attention.h:917]: why different with origin?

File: src/layers/attention.h [hunk 2]
```diff
@@ -755,60 +771,74 @@ class Attention {
...
+    template <typename AttnT>
+    void scaledDpAttention(const float *query, const AttnT *key, const AttnT *value, const float *attnMask,
+            float scale, int batchSize, int srcLen, int tgtLen, int numQHead, int numKVHead, int headSize,
+            float *output, int qStride, int kvStride, int stride) {
         // output = trans(softmax(query * trans(key)) * value)
         int nth = omp_get_max_threads();
-        int minBlk = (nth >= batchSize * numQHead ? 256 : 512);
-        int srcBlk = std::min(minBlk, srcLen);
-        int tgtBlk = std::min(minBlk, tgtLen);
+        int minBlk = (int)std::pow(2, int(std::log2(srcLen / 2)));
```
Comment (pujiang2018) [inline src/layers/attention.h:783]: any design principle here? if any, would you make some comment?

File: src/layers/attention.h [hunk 3]
```diff
@@ -755,60 +771,74 @@ class Attention {
...
+        float **thrPtrBuf;
         float **preSum = thrPtrBuf;
         float **sum = thrPtrBuf + nth;
         float **preMax = thrPtrBuf + nth * 2;
         float **max = thrPtrBuf + nth * 3;
         float **qkArr = thrPtrBuf + nth * 4;
         float **expQkvArr = thrPtrBuf + nth * 5;
+        float **qArr = thrPtrBuf + nth * 6;
+
+        thrBuf = (float *)malloc(sizeof(float) * nth * arrStride);
```
Comment (pujiang2018) [inline src/layers/attention.h:801]: Does it better to use SimpleMemPool to get the buffer? (SimpleMemPool will maintain the buffer, so next layer directly use)

File: src/layers/mlp_chatglm2.h [hunk 4]
```diff
@@ -37,37 +37,68 @@ class ChatGLM2MLP : public LlamaMLP<WeiT> {
...
-                    blockSize * sizeof(float));
-        } else {
-            const float *weightPTR = gate_upW;
-            for (int i = 0; i < hiddenSize; i++) {
-                memcpy(gateW + i * colSplit, weightPTR + ctx->splitIdx * colSplit, colSplit * sizeof(float));
-                weightPTR += intermediateSize;
-                memcpy(upW + i * colSplit, weightPTR + ctx->splitIdx * colSplit, colSplit * sizeof(float));
-                weightPTR += intermediateSize;
+
+        int enable = (getenv("ENABLE_CAT_MLP") ? atoi(getenv("ENABLE_CAT_MLP")) : 1);
```
Comment (pujiang2018) [inline src/layers/mlp_chatglm2.h:41]: why we provide the option to disable this?

**Comment by pujiang2018** [comment]
@abenmao Have we checked the output of this PR for long tokens?

File: src/utils/decoder_util.h [hunk 5]
```diff
@@ -500,98 +502,123 @@ class DecoderUtil {
...
-#pragma omp parallel for collapse(2)
-        for (int i = 0; i < batchSize; i++) {
-            for (int k = 0; k < seqLen; k++) {
-                int srcOffEachHead = k * headSize;
-                int dstOffEachLine = k * dstStride;
-
-                for (int j = 0; j < headNum; j++) {
-                    int srcOffEachLine = j * seqLen * headSize;
-                    int dstOffEachHead = j * headSize;
+    static inline __m512 dilExpKernel(__m512 vecSrc) {
```
[the remark this answers is not in the document]
Comment (abenmao) [reply src/utils/decoder_util.h]: Have changed to our vexp func

File: src/layers/attention.h (hunk 1 above)
Comment (abenmao) [reply src/layers/attention.h]: It is just for performance tuning. Forgot to rollback. Already add comments.

File: src/layers/attention.h (hunk 2 above)
Comment (abenmao) [reply src/layers/attention.h:783]: The current block size is just derived from practical experience. Added comments.

**Comment by abenmao** [comment]
> @abenmao Have we checked the output of this PR for long tokens?

Yes, have checked the output of long prompt for those models

File: src/layers/mlp_chatglm2.h (hunk 4 above)
Comment (abenmao) [reply src/layers/mlp_chatglm2.h:41]: Leave the option temporarily for future performance tuning. Maybe we can delete it after a while.

Comment (pujiang2018) [reply src/layers/mlp_chatglm2.h:41]: maybe a better method is to use a static variable, thus do not need to call 'getenv' every time.
(and if it is used multiple times in the file, we may declare a global static variable)

File: src/layers/mlp_llama.h [hunk 6]
```diff
@@ -114,25 +130,41 @@ class LlamaMLP : public SingletonBase<LlamaMLP<WeiT>> {
...
+        if (enable == 0) {
+            auto &imBuffer = ctx->imOut;
+            gateProj(doLnBefore ? normBuffer : inBuffer, imBuffer);
 
 #ifdef DEBUG
-        dbg.debugPrint("gateWeight:\n");
-        dbg.dumpMatrix(gateWeight);
-        dbg.debugPrint("gate output:\n");
-        dbg.dumpMatrix(imBuffer);
+            dbg.debugPrint("gateWeight:\n");
```
Comment (pujiang2018) [inline src/layers/mlp_llama.h:139]: need to format?

Comment (pujiang2018) [reply src/layers/mlp_llama.h:139]: looks like I made a mistake. The code already formatted.

File: src/layers/mlp_chatglm2.h (hunk 4 above)
Comment (abenmao) [reply src/layers/mlp_chatglm2.h]: I added global vars in mlp_llama.cpp. Please review~

**Review by pujiang2018 — APPROVED** [review]

**Comment by pujiang2018** [comment]
will merge after the build server checking.

**Pull request merged**